### PR TITLE
Code Refactoring across extension functionalities

### PR DESCRIPTION
**Pull request by sharhp**
Major Code Refactoring:
May have some elements from other recent commits on other recent branches - please merge this after pulling other pending requests (#35 and #36) for minimum conflicts.
Documented sources.
Refactored necessary source lines.
Removed unused Lines of Code.

### CODE REVIEW

File: ai-compile/vscode-extension/src/codeActionProvider.ts [hunk 1]
```diff
@@ -25,6 +23,9 @@ export class EduCodeActionProvider implements vscode.CodeActionProvider {
   public createCodeActions: boolean = true;
   public feedbackLevel = FeedbackLevel.novice;
 
+  /* This function helps modify the default range for a fix returned by pymacer.
```
Comment (annafariha) [inline ai-compile/vscode-extension/src/codeActionProvider.ts:26]: This comment is not clear enough. What is "Default range"?

File: ai-compile/vscode-extension/src/codeActionProvider.ts [hunk 2]
```diff
@@ -37,16 +38,17 @@ export class EduCodeActionProvider implements vscode.CodeActionProvider {
...
     lineText = lineText.slice(edit.start);
-    const startCharCode = lineText.charCodeAt(0);
+    const lineStartCharCode = lineText.charCodeAt(0);
     
-    // underline next word (in case of a starting WS character - typical of indentation fixes)
-    if(startCharCode === " ".charCodeAt(0)) {
+    // capture range of next word (in case of a starting whitespace (WS) character - typical of indentation fixes)
+    if(lineStartCharCode === " ".charCodeAt(0)) {
       
-      const alNumRegExp = /[a-z0-9_]+/ig;
```
Comment (annafariha) [inline ai-compile/vscode-extension/src/codeActionProvider.ts:45]: Use some built-in library for this and do not reimplement checking if something is alphanumeric.

File: ai-compile/vscode-extension/src/decorator.ts [hunk 3]
```diff
@@ -179,7 +159,7 @@ export class Decorator {
       });
     }
 
-    if (diagnosticLevel < 3) {
+    if (activeHighlight < 2) {
```
Comment (annafariha) [inline ai-compile/vscode-extension/src/decorator.ts]: Use a constant to assign value 2/3, never hard-code values in code like this.

File: ai-compile/vscode-extension/src/codeActionProvider.ts (hunk 1 above)
Comment (sharhp) [reply ai-compile/vscode-extension/src/codeActionProvider.ts:26]: Please refer to the Edit interface under pymacer.ts - hope that would clarify this.

File: ai-compile/vscode-extension/src/pymacer.ts [hunk 4]
```diff
@@ -1,80 +1,108 @@
...
+  public set filePath(value: string) {
+    this._filePath = value;
+  }
+}
+
+export const serverURL: string | undefined = vscode.workspace
   .getConfiguration("python-hints")
   .get("webServerPath");
 
 export const requestTimeOut: number = 100000;
```
Comment (annafariha) [inline ai-compile/vscode-extension/src/pymacer.ts]: Have a new file called `Config` and pull all these hard-coded numbers from there. Again, never hardcode any constant within the source code.

File: ai-compile/vscode-extension/src/codeActionProvider.ts (hunk 2 above)
Comment (sharhp) [reply ai-compile/vscode-extension/src/codeActionProvider.ts:45]: Could you please point us to any built-in library for achieving this? Searched for it online and couldn't really find one, that's probably why we couldn't include it in the first place.
There are 3rd party libraries however, which isn't much different from the above implementation, given the trivial nature of the implementation.

**Pull request merged**